Include/datetime.h

```diff
@@ -224,6 +207,8 @@ static PyDateTime_CAPI *PyDateTimeAPI = NULL;
 
 #define PyTZInfo_Check(op) PyObject_TypeCheck(op, PyDateTimeAPI->TZInfoType)
```

Comment:
Why these macros are defined for non-core C API at all? `PyDate_Check()` and like are not defined.

Reply:
I don't understand what you mean, `PyDate_Check` is defined in this block, these are all part of the [documented C API](https://docs.python.org/3/c-api/datetime.html).

Reply:
I don't understand. In which case PyDate_Check() is not defined when you include datetime.h?

Reply:
Ah, sorry, I was confused by changes in this PR. Now `PyDate_Check` is defined unconditionally, but `PyTZInfo_Check` only if `_PY_DATETIME_IMPL` is not defined, right? Then the question is why it is.

Reply:
Before Victor's PR, all of these macros are defined here, twice - they have different definitions when built as part of CPython than when built as part of an extension module.

The reason they are different is that the datetime module exposes all of its functionality via a capsule (`PyDateTimeAPI`), but the datetime module itself doesn't need to use the capsule because it has direct access to the relevant symbols.

Victor's PR takes the "CPython only" builds and moves them out of `datetime.h` and into `_datetimemodule.c`, which is the only place they are (and can be) used. The macros are under `_PY_DATETIME_IMPL` in this case so that he doesn't have to `#undef` them in `_datetimemodules.c`.

There's probably a case to be made in favor of dropping the macros defined in `_datetimemodules.c` *entirely* and having `_datetimemodules.c` just use the public macros, but I am not sure what the consequences there could be. I imagine a performance hit for starters (probably minor), and it would mean that `_datetimemodules.c` would need access to the `PyDateTimeAPI` singleton as well.

I think Victor's current proposed changes are an improvement over the current state of things and unless we missed something, should have no change in behavior for end users while making `datetime` easier to maintain. We can start with these changes and then do an analysis later of whether the code can be further improved by unifying the `*_Check` macros and having `_datetimemodule.c` generally use the public API where possible.

Reply:
Also, hopefully I understood your comment correctly. If you meant that `PyTZInfo_Check` is different from the other ones, I don't think it is. `PyTZInfo_Check` is used in [`_datetimemodule.c](https://github.com/python/cpython/blob/master/Modules/_datetimemodule.c#L1111), so it has to be present in both the public and private APIs.

Reply:
I would add that currently, including datetime.h in a C file which defines Py_BUILD_CORE doesn't work, since it uses types which are only exported by an external library... (_datetime is compared as a shared library by default on Linux)